fix: allow np.scalar to be used in Series.__getitem__

thevro opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [X] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- closes #1493

## Checklist

- [X] Code follows style guide (ruff)
- [ ] Tests added
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below
```python
>>> import pandas as pd
>>> import narwhals as nw
>>> s = pd.Series([0,1,2])
>>> snw = nw.from_native(s, series_only=True)
>>> snw[snw.min()]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/git/narwhals/narwhals/series.py", line 72, in __getitem__
    return self._from_compliant_series(self._compliant_series[idx])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/git/narwhals/narwhals/series.py", line 243, in _from_compliant_series
    return self.__class__(
           ^^^^^^^^^^^^^^^
  File "/git/narwhals/narwhals/series.py", line 58, in __init__
    raise AssertionError(msg)
AssertionError: Expected Polars Series or an object which implements `__narwhals_series__`, got: <class 'numpy.int64'>.
```
Previous error was an `AttributeError` arising from `_pandas_like/series.py`.

MarcoGorelli reviewed (commented):
thanks @thevro , can you add a test please?

FBruzzesi commented on narwhals/_pandas_like/series.py:
```diff
@@ -112,7 +113,8 @@ def __getitem__(self, idx: int) -> Any: ...
 
     def __getitem__(self, idx: int | slice | Sequence[int]) -> Any | Self:
-        if isinstance(idx, int):
+        np = get_numpy()
+        if np.isscalar(idx):
```
Should this be a one liner to also check numpy is installed?

```suggestion
        if isisntance(idx, int) or ((np := get_numpy()) is not None and np.isscalar(idx)):
```

thevro replied:
This is pandas-specific code, so isn't numpy guaranteed to be installed? Still, there's no harm checking.
You have a typo: could you please change `isisntance` to `isinstance`? Then I will commit your suggestion.
If you know of a more general way to check for scalars, that might be preferable.

FBruzzesi replied:
Yeah sorry, I realized that after the comment, apology for the inconvenience! I am resolving the comment

FBruzzesi reviewed (commented):
Hey I left a tiny comment. I am concerned that if numpy is not installed then this could fail

MarcoGorelli commented:
thanks @thevro , looks like there's some test failures

thevro commented:
> thanks @thevro , looks like there's some test failures

The test that fails is the one I added to confirm the exception demonstrated in the PR description. Instead of raising an AttributeError in `narwhals/_pandas_like/series.py`, as described in #1493, we now raise an AssertionError in `narwhals/series.py`. Should I change the test so that it expects the new exception (which is still a bug), or shall we try to pass the test in its current form within this PR?

Here is the reason we still get an error:
Yesterday I initially applied the same approach of `isinstance`-comparing to `numbers.Number` instead of to `int` in _both_ files, and that caused code equivalent to the test I added today to pass. However, it occurred to me that pd.Series can be indexed by other scalars such as strings (scalar != non-Iterable, sadly), so a `np.isscalar` approach is more appropriate. But `narwhals/series.py` must work without numpy installed. I see two possible fixes:

1. if `__getitem__` in `narwhals/_pandas_like/series.py` receives a scalar, the code in `narwhals/series.py` shouldn't even be called.
2. `narwhals/series.py` should also check for scalars in general instead of just `int`s. We would need to find a way to do this without using numpy.

MarcoGorelli commented:
I think you could just make a function in `narwhals/dependencies.py` called `is_numpy_scalar` in which you just do
```
return (np := get_numpy()) is not None and np.isscalar(obj)
```

Like this, it checks if `obj` is a NumPy scalar without importing NumPy. Like this, you can safely use it in `narwhals/series.py`

MarcoGorelli commented on tests/series_only/scalar_index_test.py:
```diff
@@ -0,0 +1,12 @@
+
+import narwhals.stable.v1 as nw
+
+np = nw.dependencies.get_numpy()
+pd = nw.dependencies.get_pandas()
```
you can just import these directly here (though if you use `constructor_eager` they're probably not necessary)

thevro replied:
Thank you. I can confirm that `constructor_eager` works without them.

MarcoGorelli reviewed (commented):
thanks @thevro , this is close

thevro commented on narwhals/dependencies.py:
```diff
@@ -220,6 +220,11 @@ def is_numpy_array(arr: Any) -> TypeGuard[np.ndarray]:
 
 
+def is_numpy_scalar(scalar: Any) -> TypeGuard[np.generic]:
+    """Check whether `scalar` is a NumPy Array without importing NumPy."""
```
Thank you. See https://github.com/narwhals-dev/narwhals/pull/1515/commits/7a5aa74dd6ef3f63b765bfe14d9594c2bb9b1eab

thevro commented on tests/series_only/scalar_index_test.py:
```diff
@@ -0,0 +1,12 @@
+pd = nw.dependencies.get_pandas()
+
+
+def test_index() -> None:
+    s = pd.Series([0, 1, 2])
+    snw = nw.from_native(s, series_only=True)
+    assert snw[snw[0]] == np.int64(0)
```
I didn't understand `constructor_eager` before, but having tried it out, it seems like it runs the test for different libraries. Correct?
The problem here is that, while `numpy.int64(0) == 0`, for some reason `<pyarrow.Int64Scalar: 0> != 0`. Not sure about other libraries...

MarcoGorelli replied:
thanks @thevro ! yeah `Series.__getitem__` should return a Python scalar for PyArrow. i'll just make a PR, then if you fetch and merge then the comparison should work

MarcoGorelli replied:
done https://github.com/narwhals-dev/narwhals/pull/1554

if you fetch upstream and then merge upstream/main then the `== 0` comparison should work

thevro replied:
And doing
```python
import pyarrow as pa
from pyarrow import compute as pc
pazero = pc.cast(0, pa.int64())
...
idx = snw[0]  # test indexing using builtin int
assert idx == 0 or idx == pazero # idx should have type numpy.int64 or pa.int64 
assert snw[idx] == 0 or idx == pazero # test indexing using third-party int
```
produces a test error:
```python
>       assert snw[idx] == 0 or idx == pazero # test indexing using numpy.int64

tests/series_only/scalar_index_test.py:19: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
narwhals/series.py:73: in __getitem__
    return self._from_compliant_series(self._compliant_series[idx])
narwhals/_arrow/series.py:378: in __getitem__
    return self._from_native_series(self._native_series[idx])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: 'pyarrow.lib.Int64Scalar' object cannot be interpreted as an integer

pyarrow/table.pxi:312: TypeError
```
I suppose this comes of handling only numpy scalars.

MarcoGorelli replied:
yup, fixed now, if you fetch and merge upstream/main then `== 0` should work

thevro commented on narwhals/series.py:
```diff
@@ -67,7 +68,7 @@ def __getitem__(self: Self, idx: int) -> Any: ...
 
     def __getitem__(self: Self, idx: int | slice | Sequence[int]) -> Any | Self:
-        if isinstance(idx, int):
+        if isinstance(idx, int) or is_numpy_scalar(idx):
```
As I understand it, we would ideally capture any sort of numpy scalar that can be used to index a pandas series. Can't strings, floats and datetime objects also be used?

MarcoGorelli replied:
no, we only allow positional indexing here

MarcoGorelli reviewed (approved):
thanks @thevro 

![](https://i.fluffy.cc/4HrdmGw7MHL3rcH3J0lCBBG7xhlxnWs7.gif)
